x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -692,6 +704,29 @@ GetUserPrivilegesResponse buildUserPrivilegesResponseObject(Role userRole) {
         return Collections.unmodifiableSet(indicesAndAliases);
     }
 
+    static Predicate<IndexAbstraction> buildPredicateForAuthorizedIndicesFromRole(Role role, RequestInfo requestInfo, Metadata metadata) {
+        final Predicate<IndexAbstraction> predicate = role.allowedIndicesMatcher(requestInfo.getAction());
+
+        // do not include data streams for actions that do not operate on data streams
+        TransportRequest request = requestInfo.getRequest();
+        final boolean includeDataStreams = (request instanceof IndicesRequest) && ((IndicesRequest) request).includeDataStreams();
+
+        return indexAbstraction -> {
+            if (indexAbstraction == null) {
+                return false;
+            }
+            if (includeDataStreams) {
+                if (predicate.test(indexAbstraction)) {
+                    return true;
+                } else {
+                    return indexAbstraction.getParentDataStream() != null && predicate.test(indexAbstraction.getParentDataStream());
+                }
+            } else {
+                return indexAbstraction.getType() != IndexAbstraction.Type.DATA_STREAM && predicate.test(indexAbstraction);
+            }
+        };
+    }
```

Comment:
This predicate when evaluated over all the names in the cluster state will end-up being evaluated twice over data stream names, no?
But maybe it's not a problem?

Reply:
You are right. I didn't think of it. When privilege is _not_ granted for backing index names (the most probable case), running a backing index name through the predicate ends up always evaluating both the index name and the data stream name.

I don't know how bad it is. But definitely _not_ ideal. It is a bit trickly to solve. I came up with following solution:

1. Reversed the order of testing of index name and data stream name. It now always check data stream name first. This should be better for testing concrete requested index names because (1) The requested indices are most likely data streams instead of backing indices; (2) Privileges are likely granted over data streams. Therefore, when the concrete index name should be authorized, it will likely be a single test with data stream name or a single test with the concrete index name (if it does not belong to a data stream). When it should not be authorized, both tests can run. But this will be a 403 error case, performance is less of an issue here. 

2. NOTE the 1st point does not help when we must evaluated through all indices, i.e. when wildcard must be expanded. In this case, the caller of the predicate nows does its own optimisation to add all backing indices of a data stream if the data stream is authorized. This avoids running the predicate for backing index names.